packages/@ember/-internals/runtime/lib/mixins/array.d.ts

```diff
@@ -0,0 +1,103 @@
+type Value<T, K extends string> = K extends keyof T ? T[K] : unknown;
+
+export type NatifiedArray<T> = T[] & NativeArray<T>;
+
+interface EmberArray<T> {
+  length: number;
+  objectAt(idx: number): T | undefined;
+}
+declare class EmberArray<T> {
+  objectsAt(indexes: number[]): Array<T | undefined>;
+  get firstObject(): T | undefined;
+  get lastObject(): T | undefined;
+  slice(beginIndex?: number, endIndex?: number): NatifiedArray<T>;
+  indexOf(object: T, startAt?: number): number;
+  lastIndexOf(object: T, startAt?: number): number;
+  forEach<Target>(
+    callback: (this: Target, item: T, index: number, arr: this) => void,
+    target?: Target
+  ): this;
+  getEach<K extends string>(key: K): NatifiedArray<Value<T, K>>;
+  setEach<K extends string>(key: K, value: Value<T, K>): this;
+  map<R, Target>(
+    callback: (this: Target, item: T, index: number, arr: this) => R,
+    target?: Target
+  ): NatifiedArray<R>;
```

Comment:
Unless we have a good reason to, we should make this match the `map` from JS `Array`—either manually, or by simply doing `Array<T>['map']` (which is what we do on DefinitelyTyped):

```suggestion
  map: Array<T>['map'];
```

While I am sympathetic to the interest of being more robust than TS' own types are, in our experience on Typed Ember that has actually caused more difficulty (e.g. because it makes assignability more complicated, etc.).

---

That being said: if we know that this returns `NativeArray` rather than `EmberArray`, this makes a fair bit of sense. In that case, it's conventional to use `U` for mapping functions (T → U).

```suggestion
  map<U, Target>(
    callback: (this: Target, item: T, index: number, arr: this) => R,
    target?: Target
  ): NativeArray<U>;
```

Reply:
Yes, it does return a NativeArray, that said, it seems like the true native Array types might win out in TS anyway.

Reply:
I figured out why the native types were winning and resolved that.